src/main/scala/is/hail/expr/Parser.scala

```diff
@@ -819,11 +819,11 @@ object Parser extends JavaTokenParsers {
       } |
       "MatrixAggregateColsByKey" ~> matrix_ir ~ ir_value_expr ^^ { case child ~ agg => ir.MatrixAggregateColsByKey(child, agg) } |
       "MatrixAggregateRowsByKey" ~> matrix_ir ~ ir_value_expr ^^ { case child ~ agg => ir.MatrixAggregateRowsByKey(child, agg) } |
-      "MatrixRange" ~> int32_literal ~ int32_literal ~ int32_literal_opt ~ boolean_literal ~ boolean_literal ^^ { case nRows ~ nCols ~ nPartitions ~ dropCols ~ dropRows =>
-        MatrixIR.range(HailContext.get, nRows, nCols, nPartitions, dropCols, dropRows)
-      } |
-      "MatrixRead" ~> string_literal ~ boolean_literal ~ boolean_literal ~ matrix_type_expr_opt ^^ { case path ~ dropCols ~ dropRows ~ requestedType =>
-        MatrixIR.read(HailContext.get, path, dropCols, dropRows, requestedType)
+      "MatrixRead" ~> matrix_type_expr_opt ~ boolean_literal ~ boolean_literal ~ string_literal ^^ {
+        case typ ~ dropCols ~ dropRows ~ readerStr =>
+          implicit val formats = ir.MatrixReader.formats
+          val reader = Serialization.read[ir.MatrixReader](readerStr)
+          ir.MatrixRead(typ, dropCols, dropRows, reader)
```

Comment:
I personally would make MatrixRead take a non-optional type and pass in `typ.getOrElse(reader.typ)` here, I think that's a tad simpler.  But this is also fine.